pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,746 @@
+
+		ensure!(signature.verify(&wrapped_payload[..], &key), Error::<T>::InvalidSignature);
+
+		Ok(())
+	}
+
+	/// The furthest in the future a mortality_block value is allowed
+	/// to be for current_block
+	/// This is calculated to be past the risk of a replay attack
+	fn mortality_block_limit(current_block: T::BlockNumber) -> T::BlockNumber {
+		current_block + T::BlockNumber::from(T::MortalityWindowSize::get())
+	}
+
+	fn check_schema(
+		schema_id: SchemaId,
+		expected_payload_location: PayloadLocation,
+		is_payload_signed: bool,
+		is_deleting: bool,
+	) -> DispatchResult {
+		let schema =
+			T::SchemaProvider::get_schema_by_id(schema_id).ok_or(Error::<T>::InvalidSchemaId)?;
+
+		// Ensure that the schema's payload location matches the expected location.
+		ensure!(
+			schema.payload_location == expected_payload_location,
+			Error::<T>::SchemaPayloadLocationMismatch
+		);
+
+		// Ensure that the schema allows signed payloads.
+		// If so, calling extrinsic must be of signature type.
+		if schema.settings.contains(&SchemaSetting::SignatureRequired) {
+			ensure!(is_payload_signed, Error::<T>::SchemaNotSupported);
+		}
+
+		// Ensure that the schema does not allow deletion for AppendOnly SchemaSetting.
+		if schema.settings.contains(&SchemaSetting::AppendOnly) {
+			ensure!(!is_deleting, Error::<T>::SchemaNotSupported);
+		}
+
+		Ok(())
+	}
+
+	fn check_grants(
+		provider_key: T::AccountId,
+		state_owner_msa_id: MessageSourceId,
+		schema_id: SchemaId,
+	) -> Result<MessageSourceId, DispatchError> {
+		let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
+			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
+
+		// if provider and owner are the same no delegation is needed
+		if provider_msa_id != state_owner_msa_id {
+			let current_block = frame_system::Pallet::<T>::block_number();
+			T::SchemaGrantValidator::ensure_valid_schema_grant(
+				ProviderId(provider_msa_id),
+				DelegatorId(state_owner_msa_id),
+				schema_id,
+				current_block,
+			)
+			.map_err(|_| Error::<T>::UnAuthorizedDelegate)?;
+		}
+
+		Ok(provider_msa_id)
+	}
+
+	/// Verifies if the key has an Msa and if it matches with expected one
+	///
+	/// # Errors
+	/// * [`Error::InvalidMessageSourceAccount`]
+	///
+	fn check_msa(key: T::AccountId, expected_msa_id: MessageSourceId) -> DispatchResult {
+		let state_owner_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&key)
+			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
+		ensure!(state_owner_msa_id == expected_msa_id, Error::<T>::InvalidMessageSourceAccount);
+		Ok(())
+	}
+
+	/// Verifies the actions size
+	///
+	/// # Errors
+	/// * [`Error::ItemExceedsMaxBlobSizeBytes`]
+	///
+	fn check_actions(
+		actions: &BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
+	) -> DispatchResult {
+		ensure!(
+			actions.iter().all(|a| match a {
+				ItemAction::Add { data } =>
+					data.len() <= T::MaxItemizedBlobSizeBytes::get() as usize,
```

Comment:
is this the size check I was looking for?

Reply:
Correct but I believe we should also reflect that in our ds using BoundedVec over page data size limit

Reply:
@saraswatpuneet  I'm not sure if I understand what you mean but we check the page size whenever we are applying those actions